### PR TITLE
Refactor reading routine with potential buffer overflow

### PR DESCRIPTION
**Pull request by aradi**
Code produced by the NAG compiler was crashing at this point

### CODE REVIEW

File: src/dftbp/solvation/gbsafile.F90 [hunk 1]
```diff
@@ -216,62 +219,26 @@ subroutine nextLine(unit, line, lineno, file, iostat, node)
...
 
-    read(unit, '(a)', advance='no', iostat=iErr, iomsg=iomsg, size=length) line
-    if (length >= len(line)) then
+    if (present(iostat)) then
+      iostat = iErr
+    else if (iErr /= 0) then
       if (present(file)) then
-        write(errorStr, '(a, "(", i0, "):", 1x, a)') &
-            & trim(file), lineno, "too many characters, line truncated"
+        write(errorStr, '(a, "(", i0, "):", 1x, a)') trim(file), lineno,&
```
Comment (bhourahine) [inline src/dftbp/solvation/gbsafile.F90:232]: How about a more informative message, something like:
While reading file XXX, an error was encountered on line YYY
?

File: src/dftbp/solvation/gbsafile.F90 [hunk 2]
```diff
@@ -216,62 +219,26 @@ subroutine nextLine(unit, line, lineno, file, iostat, node)
...
+            & "encountered error while reading line"
       else
         write(errorStr, '(a, 1x, i0, ":", 1x, a)') &
-            & "line", lineno, "too many characters, line truncated"
+           & "line", lineno, "encountered error while reading line"
```
Comment (bhourahine) [inline src/dftbp/solvation/gbsafile.F90:236]: Likewise:
Encountered an error when reading line YYY

File: src/dftbp/solvation/gbsafile.F90 [hunk 3]
```diff
@@ -216,62 +219,26 @@ subroutine nextLine(unit, line, lineno, file, iostat, node)
...
 
     lineno = lineno + 1
+    call lineReader%readLine(line, iErr)
 
-    read(unit, '(a)', advance='no', iostat=iErr, iomsg=iomsg, size=length) line
-    if (length >= len(line)) then
+    if (present(iostat)) then
+      iostat = iErr
+    else if (iErr /= 0) then
       if (present(file)) then
```
Comment (bhourahine) [inline src/dftbp/solvation/gbsafile.F90:231]: Am I missing something, or is file always present in calls to this routine?

**Review by bhourahine — COMMENTED** [review]
Seems correct, but not sure if there is anything weird about the parameter files.

File: src/dftbp/solvation/gbsafile.F90 (hunk 3 above)
Comment (aradi) [reply src/dftbp/solvation/gbsafile.F90]: Fair enough. I've reduced the routine now to the bare minimum according to the actual use cases.

**Comment by aradi** [comment]
Should be fine now. Changed error message and eliminated all superfluous code path in `nextLine()`

**Review by bhourahine — APPROVED** [review]
Other than updating and possibly squashing, could you add a comment in the commit log to the effect that the nexline routine now only reads from a named file, as was never called with a unit.

**Pull request merged**